# 1
Import the tools.

In [1]:
import pandas as pd
import numpy as np
import copy

# 2
Read the data.

In [2]:
flights = pd.read_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dati\\flights.csv')
sectorhours = pd.read_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dati\\sectorhours.csv')
trajectories = pd.read_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dati\\trajectories.csv')

# 3
Print the data.

In [3]:
#print(flights)
#print (sectorhours)
#print (trajectories)
flights

uuid           O_D_A             dep_time  \
0      GRL780_BGSF_EKCH_20170901134000  BGSF_EKCH_A332  2017-09-01 13:50:00   
1      N174JF_BGSF_LOWI_20170901103500  BGSF_LOWI_AT43  2017-09-01 10:45:00   
2      WOW462_BIKF_EBBR_20170901065500  BIKF_EBBR_A321  2017-09-01 06:25:00   
3      ICE554_BIKF_EBBR_20170901074000  BIKF_EBBR_B752  2017-09-01 07:50:00   
4      WOW720_BIKF_EDDB_20170901054500  BIKF_EDDB_A321  2017-09-01 05:55:00   
...                                ...             ...                  ...   
29912   HVN31_VVTS_EDDF_20170901155500  VVTS_EDDF_A332  2017-09-01 16:05:00   
29913   HVN51_VVTS_EGLL_20170901173500  VVTS_EGLL_A332  2017-09-01 17:45:00   
29914  AFR253_VVTS_LFPG_20170901013000  VVTS_LFPG_A332  2017-09-01 01:40:00   
29915  THY166_VVTS_LTBA_20170901144000  VVTS_LTBA_A332  2017-09-01 14:50:00   
29916  BWJ998_ZYHB_EGGW_20170901110000  ZYHB_EGGW_E190  2017-09-01 11:10:00   

      original_trajectory  dep_time_minutes  
0        BGSF_EKCH_A332_3               830  
1        BGSF_LOWI_AT43_1               645  
2        BIKF_EBBR_A321_1               385  
3        BIKF_EBBR_B752_1               470  
4        BIKF_EDDB_A321_1               355  
...                   ...               ...  
29912    VVTS_EDDF_A332_2               965  
29913    VVTS_EGLL_A332_3              1065  
29914    VVTS_LFPG_A332_3               100  
29915    VVTS_LTBA_A332_1               890  
29916    ZYHB_EGGW_E190_1               670  

[29917 rows x 5 columns]

# 4
Drop unnecessary columns in "flights".

In [4]:
flights = flights.drop(columns=['original_trajectory', 'dep_time'])
flights

uuid           O_D_A  dep_time_minutes
0      GRL780_BGSF_EKCH_20170901134000  BGSF_EKCH_A332               830
1      N174JF_BGSF_LOWI_20170901103500  BGSF_LOWI_AT43               645
2      WOW462_BIKF_EBBR_20170901065500  BIKF_EBBR_A321               385
3      ICE554_BIKF_EBBR_20170901074000  BIKF_EBBR_B752               470
4      WOW720_BIKF_EDDB_20170901054500  BIKF_EDDB_A321               355
...                                ...             ...               ...
29912   HVN31_VVTS_EDDF_20170901155500  VVTS_EDDF_A332               965
29913   HVN51_VVTS_EGLL_20170901173500  VVTS_EGLL_A332              1065
29914  AFR253_VVTS_LFPG_20170901013000  VVTS_LFPG_A332               100
29915  THY166_VVTS_LTBA_20170901144000  VVTS_LTBA_A332               890
29916  BWJ998_ZYHB_EGGW_20170901110000  ZYHB_EGGW_E190               670

[29917 rows x 3 columns]

# 5
Consider only the sectorhours with role 'E'. <br>Reset the index.

In [5]:
sectorhours = sectorhours[sectorhours['role']=='E']
sectorhours.reset_index(drop=True, inplace=True)
sectorhours

sector_hours      icao role  minute_start  minute_end  capacity
0       EGTTOCK_0000_0100_E   EGTTOCK    E             0          59        49
1       EGTTOCK_0100_0200_E   EGTTOCK    E            60         119        49
2       EGTTOCK_0200_0300_E   EGTTOCK    E           120         179        49
3       EGTTOCK_0300_0400_E   EGTTOCK    E           180         239        49
4       EGTTOCK_0400_0500_E   EGTTOCK    E           240         299        49
...                     ...       ...  ...           ...         ...       ...
13485   LSAGL67_2010_2040_E   LSAGL67    E          1210        1239        23
13486     LYBAL_1659_1800_E     LYBAL    E          1019        1079        41
13487    LYBATN_1600_1700_E    LYBATN    E           960        1019        43
13488    LYBAUN_1600_1700_E    LYBAUN    E           960        1019        43
13489  LYBAUWES_1659_1800_E  LYBAUWES    E          1019        1079        54

[13490 rows x 6 columns]

# 6
Create a new column 'O_D_A' in trajectories, obtained from the first 14 characters of 'centroid'.
<br> Drop the unnecessary columns.
<br> Rename the columnns.
<br> Check that the entries of 'O_D_A' are all of lenght 14. 

In [6]:
trajectories['O_D_A'] = trajectories['centroid'].str[0:14]
trajectories = trajectories.drop(columns=['eobt', 'time_entry'])
trajectories = trajectories[['O_D_A', 'centroid', 'icao', 'time_to_sector_from_origin']]

In [7]:
print(len(trajectories))
print(len(trajectories[(trajectories['O_D_A']).str.len()==14]))
print(len(trajectories[(trajectories['O_D_A']).str.len()!=14]))
trajectories

2466877
2466877
0


O_D_A          centroid      icao  \
0        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKLVJEWH   
1        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVO355   
2        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2    UKBVSO   
3        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   UKBVTC1   
4        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVTC12   
...                 ...               ...       ...   
2466872  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGPSA   
2466873  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGADF   
2466874  LMML_HLLB_OTHR  LMML_HLLB_OTHR_1   LMMMWST   
2466875  LMML_HLLB_OTHR  LMML_HLLB_OTHR_1   LMMMALL   
2466876  LMML_HLLB_OTHR  LMML_HLLB_OTHR_1   HLLLALL   

         time_to_sector_from_origin  
0                                20  
1                                53  
2                                53  
3                                73  
4                                73  
...                             ...  
2466872                         112  
2466873                         113  
2466874                          26  
2466875                          26  
2466876                          54  

[2466877 rows x 4 columns]

# 7
Merge "trajectories" with "flights".

E' indifferente applicare merge con modalità 'inner' o 'right': tutte le traiettorie in flights hanno un corrispondente in trajectories (il viceversa non è vero, come si potrebbe verificare applicando la modalità 'left': ci sono traiettorie in trajectories che non sono state volate il 1/9/2017).

In [8]:
merged = trajectories.merge(flights, how='right', on = 'O_D_A')
merged

O_D_A          centroid      icao  \
0        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKLVJEWH   
1        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVO355   
2        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2    UKBVSO   
3        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   UKBVTC1   
4        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVTC12   
...                 ...               ...       ...   
4135268  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGKTG   
4135269  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1     EDGG4   
4135270  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1  EDGGKIHA   
4135271  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGPSA   
4135272  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGADF   

         time_to_sector_from_origin                             uuid  \
0                                20  WRC124_UKLL_UKDD_20170901180000   
1                                53  WRC124_UKLL_UKDD_20170901180000   
2                                53  WRC124_UKLL_UKDD_20170901180000   
3                                73  WRC124_UKLL_UKDD_20170901180000   
4                                73  WRC124_UKLL_UKDD_20170901180000   
...                             ...                              ...   
4135268                         104   Z3MKD_LWSK_EDDF_20170901130000   
4135269                         104   Z3MKD_LWSK_EDDF_20170901130000   
4135270                         104   Z3MKD_LWSK_EDDF_20170901130000   
4135271                         112   Z3MKD_LWSK_EDDF_20170901130000   
4135272                         113   Z3MKD_LWSK_EDDF_20170901130000   

         dep_time_minutes  
0                    1090  
1                    1090  
2                    1090  
3                    1090  
4                    1090  
...                   ...  
4135268               785  
4135269               785  
4135270               785  
4135271               785  
4135272               785  

[4135273 rows x 6 columns]

# 8
Drop unnecessary columns.
<br>Reorder the columns.
<br>Rename the columns.

In [9]:
merged = merged[['uuid', 'O_D_A', 'centroid', 'icao', 'dep_time_minutes', 'time_to_sector_from_origin']]
merged.columns = ['flight', 'O_D_A', 'trajectory', 'icao', 'dep_time', 'time_to_sector']
merged

flight           O_D_A        trajectory  \
0        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                  ...             ...               ...   
4135268   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135269   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135270   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135271   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135272   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

             icao  dep_time  time_to_sector  
0        UKLVJEWH      1090              20  
1        UKBVO355      1090              53  
2          UKBVSO      1090              53  
3         UKBVTC1      1090              73  
4        UKBVTC12      1090              73  
...           ...       ...             ...  
4135268   EDGGKTG       785             104  
4135269     EDGG4       785             104  
4135270  EDGGKIHA       785             104  
4135271   EDGGPSA       785             112  
4135272   EDGGADF       785             113  

[4135273 rows x 6 columns]

# 9
Create a new column 'entry', describing the time a flight enters a sector-hour.
<br>The modulo operation is needed since we might exceed midnight (60x24=1440).

In [10]:
merged['entry'] = (merged['dep_time'] + merged['time_to_sector'])%1440
merged

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


flight           O_D_A        trajectory  \
0        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                  ...             ...               ...   
4135268   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135269   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135270   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135271   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135272   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

             icao  dep_time  time_to_sector  entry  
0        UKLVJEWH      1090              20   1110  
1        UKBVO355      1090              53   1143  
2          UKBVSO      1090              53   1143  
3         UKBVTC1      1090              73   1163  
4        UKBVTC12      1090              73   1163  
...           ...       ...             ...    ...  
4135268   EDGGKTG       785             104    889  
4135269     EDGG4       785             104    889  
4135270  EDGGKIHA       785             104    889  
4135271   EDGGPSA       785             112    897  
4135272   EDGGADF       785             113    898  

[4135273 rows x 7 columns]

# 10
Check that the maximum value of the column 'entry' doesn't exceed 1440.

In [11]:
print ("Is the maximum value of the column 'entry' smaller than 1440?")
print (max(merged['entry'])<1440)
print("max(merged['entry']) =", max(merged['entry']))

Is the maximum value of the column 'entry' smaller than 1440?
True
max(merged['entry']) = 1439


# 11
Merge "merged" with "sectorhours", based on 'icao'.

In [12]:
merged_2 = merged.merge(sectorhours, how='left', left_on='icao', right_on='icao')
merged_2

flight           O_D_A        trajectory  \
0         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                   ...             ...               ...   
35353263   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353264   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353265   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353266   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353267   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

              icao  dep_time  time_to_sector  entry          sector_hours  \
0         UKLVJEWH      1090              20   1110  UKLVJEWH_0000_0100_E   
1         UKLVJEWH      1090              20   1110  UKLVJEWH_0100_0200_E   
2         UKLVJEWH      1090              20   1110  UKLVJEWH_0200_0300_E   
3         UKLVJEWH      1090              20   1110  UKLVJEWH_0300_0400_E   
4         UKLVJEWH      1090              20   1110  UKLVJEWH_0400_0500_E   
...            ...       ...             ...    ...                   ...   
35353263   EDGGPSA       785             112    897   EDGGPSA_1700_1800_E   
35353264   EDGGPSA       785             112    897   EDGGPSA_1800_1900_E   
35353265   EDGGPSA       785             112    897   EDGGPSA_1900_2000_E   
35353266   EDGGPSA       785             112    897   EDGGPSA_2000_2100_E   
35353267   EDGGADF       785             113    898                   NaN   

         role  minute_start  minute_end  capacity  
0           E           0.0        59.0      39.0  
1           E          60.0       119.0      39.0  
2           E         120.0       179.0      39.0  
3           E         180.0       239.0      39.0  
4           E         240.0       299.0      39.0  
...       ...           ...         ...       ...  
35353263    E        1020.0      1079.0      38.0  
35353264    E        1080.0      1139.0      38.0  
35353265    E        1140.0      1199.0      38.0  
35353266    E        1200.0      1259.0      38.0  
35353267  NaN           NaN         NaN       NaN  

[35353268 rows x 12 columns]

The modality 'left' keeps all the sectorhours present in "merged", even when they don't have any corrispondence in "sectorhours" (the relevant entries are set to 'NaN').
The modality 'inner', instead, would keep only the sectorhours in "merged" having an icao which is also present in "sectorhours", thus leaving out a number of flights.

# 12
Keep only the sectorhours having a 'minute_start' not greater than 'entry' +30 and an 'entry' not smaller than 'minute_end' +30.
<br>Compare the number of flights included in "merged_2" and "clean" respectively.
See #18 for the reasoning.

In [13]:
clean = merged_2[(merged_2['minute_start']<=merged_2['entry']+30) & (merged_2['entry']<=merged_2['minute_end']+30)]
clean

flight           O_D_A        trajectory  \
18        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
19        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
43        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
44        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
67        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                   ...             ...               ...   
35353208   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353227   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353228   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353260   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353261   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

              icao  dep_time  time_to_sector  entry          sector_hours  \
18        UKLVJEWH      1090              20   1110  UKLVJEWH_1800_1900_E   
19        UKLVJEWH      1090              20   1110  UKLVJEWH_1900_2000_E   
43          UKBVSO      1090              53   1143    UKBVSO_1800_1900_E   
44          UKBVSO      1090              53   1143    UKBVSO_1900_2000_E   
67        UKBVTC12      1090              73   1163  UKBVTC12_1800_1900_E   
...            ...       ...             ...    ...                   ...   
35353208   EDMMALB       785              97    882   EDMMALB_1500_1600_E   
35353227   EDGGKTG       785             104    889   EDGGKTG_1400_1500_E   
35353228   EDGGKTG       785             104    889   EDGGKTG_1500_1600_E   
35353260   EDGGPSA       785             112    897   EDGGPSA_1400_1500_E   
35353261   EDGGPSA       785             112    897   EDGGPSA_1500_1600_E   

         role  minute_start  minute_end  capacity  
18          E        1080.0      1139.0      39.0  
19          E        1140.0      1199.0      39.0  
43          E        1080.0      1139.0      36.0  
44          E        1140.0      1199.0      36.0  
67          E        1080.0      1139.0      34.0  
...       ...           ...         ...       ...  
35353208    E         900.0       959.0      46.0  
35353227    E         840.0       899.0      41.0  
35353228    E         900.0       959.0      41.0  
35353260    E         840.0       899.0      38.0  
35353261    E         900.0       959.0      38.0  

[2909278 rows x 12 columns]

In [14]:
l1 = merged_2['flight'].unique()
l2 = clean['flight'].unique()
print("len(l1): ", len(l1))
print("len(l2): ", len(l2))
diff = list(set(l1) - set(l2))
print("len(diff): ", len(diff))

len(l1):  29917
len(l2):  29872
len(diff):  45


The number of flights being equal to 29872, tells us that 45 flights are left out.
This is due to the fact that the merge modality 'left' keeps all the sectorhours present in "merged", even when they don't have any corrispondence in "sectorhours"; the missing entries are then set to 'NaN', but the inequalities of step 12 leave out the rows corresponding to such entries.
The 45 missing flights are included again at a later step (#15 and 16).

# 13
Drop unnecessary columns. <br>Reset the index.

In [15]:
clean2 = clean.drop(columns=['icao', 'time_to_sector', 'entry', 'role', 'minute_start', 'minute_end', 'capacity'])
clean2.reset_index(drop=True, inplace=True)
clean2

flight           O_D_A        trajectory  \
0        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                  ...             ...               ...   
2909273   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
2909274   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
2909275   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
2909276   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
2909277   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

         dep_time          sector_hours  
0            1090  UKLVJEWH_1800_1900_E  
1            1090  UKLVJEWH_1900_2000_E  
2            1090    UKBVSO_1800_1900_E  
3            1090    UKBVSO_1900_2000_E  
4            1090  UKBVTC12_1800_1900_E  
...           ...                   ...  
2909273       785   EDMMALB_1500_1600_E  
2909274       785   EDGGKTG_1400_1500_E  
2909275       785   EDGGKTG_1500_1600_E  
2909276       785   EDGGPSA_1400_1500_E  
2909277       785   EDGGPSA_1500_1600_E  

[2909278 rows x 5 columns]

# 14
Collapse all the sectorhours associated to each flight into a list of sectorhours.
<br>A new dataframe is created, containing the two columns 'flight' and 'sector_hours'.

In [16]:
clean3 = clean2.groupby('flight')
clean4 = clean3['sector_hours'].unique()

clean5 = pd.DataFrame({'flight':clean4.index, 'sector_hours':clean4.values})

clean5

flight  \
0        A6RTS_OMDW_EGGW_20170901055500   
1       AAB249_EBKT_LFMN_20170901070500   
2       AAB249_LFMN_LIEO_20170901090500   
3       AAB25C_LOWW_UUMO_20170901143500   
4       AAB571_EBBR_EGNV_20170901050500   
...                                 ...   
29867   XLF056_LFPG_MMUN_20170901153100   
29868  XLF1451_LFML_LFPG_20170901030500   
29869   XRO546_LFMD_EGCC_20170901180000   
29870   XRO552_EHAM_LFMD_20170901150000   
29871    Z3MKD_LWSK_EDDF_20170901130000   

                                            sector_hours  
0      [LTAAAEU67_0800_0900_E, LTAAAEU67_0900_1000_E,...  
1      [EBBUWLS_0620_0700_E, EBBUWLS_0700_0800_E, EBB...  
2      [LFMMMNST_0800_0900_E, LFMMMNST_0900_1000_E, L...  
3      [LZBBL33_1400_1500_E, LZBBL33_1500_1600_E, LZB...  
4      [EBBUNWC_0530_0600_E, EBBUNWC_0600_0620_E, EBB...  
...                                                  ...  
29867  [LFFFTH_1500_1600_E, LFFFTH_1600_1700_E, LFFFU...  
29868  [LFMMMALY_0330_0400_E, LFMMMALY_0400_0430_E, L...  
29869  [LFMMSBAM_1715_1800_E, LFMMSBAM_1800_1900_E, L...  
29870  [EHAACOD_1500_1600_E, EHAACOD_1600_1700_E, EHA...  
29871  [LWSSLOW_1200_1300_E, LWSSLOW_1300_1400_E, LYB...  

[29872 rows x 2 columns]

# 15
Recover the 45 missing flights: create a new dataframe "flights2" selecting from "flights" only the rows having an 'uuid' that is present in the list "diff". <br>Drop the unnecessary columns. <br>Rename the column; add a new column named 'sector_hours', with empty lists as entries.
<br>Check that the lenght of this new dataframe is 45.

In [17]:
flights2 = flights[flights['uuid'].isin(diff)]
flights2 = flights2.drop(columns=['O_D_A', 'dep_time_minutes'])
flights2.columns = ['flight']
flights2['sector_hours'] = [[] for _ in range(len(flights2))]
    
print(len(flights2))    
flights2

45


flight sector_hours
4768     AOS54_EGBJ_EGFF_20170901092000           []
5141    EZE612_EGFF_EGOV_20170901142500           []
5142    EZE610_EGFF_EGOV_20170901063500           []
5540     EGL08_EGHH_EGTE_20170901181500           []
7874   NJE347D_EGSS_EGGW_20170901193000           []
8897    CAI040_EHEH_EHBK_20170901074000           []
8948    HNL33A_EHFE_EHKD_20170901104500           []
8949    HNL63C_EHFQ_EHKD_20170901163000           []
8950    HNL25A_EHFT_EHKD_20170901080000           []
8960    HNL05B_EHKP_EHKD_20170901170000           []
8961    HNL70A_EHLF_EHKD_20170901101500           []
8962    HNL07A_EHLF_EHKD_20170901074500           []
8963    HNL07C_EHLF_EHKD_20170901170000           []
8964    HNL05A_EHLL_EHKD_20170901114000           []
8965    NHX15B_EHLT_EHKD_20170901173000           []
8966    HHE60B_EHLX_EHKD_20170901130900           []
8967    NHX01A_EHMF_EHKD_20170901064000           []
8968    HNL63A_EHMG_EHKD_20170901063500           []
8969    HNL25C_EHMR_EHKD_20170901163000           []
8970    HNL54A_EHPG_EHKD_20170901062000           []
8971    HNL24B_EHPM_EHKD_20170901174500           []
8972    NHX01B_EHQH_EHKD_20170901160000           []
8973    NHX15A_EHQI_EHKD_20170901073000           []
14454   AOJ75D_LDVA_LQBK_20170901053000           []
17724  TJT052P_LFMD_LFML_20170901120000           []
17802  HOP15VJ_LFML_LFBO_20170901143000           []
17803  HOP17VJ_LFML_LFBO_20170901192500           []
17826   FPO759_LFML_LFMN_20170901004000           []
18008  FPO750F_LFMN_LFML_20170901190500           []
18009   RIV202_LFMN_LFML_20170901180000           []
18215   NAK401_LFOK_LFPM_20170901071500           []
18914   NAK401_LFPM_LFOK_20170901063000           []
19245    N166G_LFPT_LFPN_20170901092000           []
19514  TVS2456_LFSB_LFST_20170901110000           []
19535  GAC018Z_LFSB_LSZH_20170901081000           []
23949  AUA165Q_LOWG_LOWI_20170901200500           []
24288  AUA935T_LOWW_LOWK_20170901075500           []
24289   AUA99J_LOWW_LOWK_20170901151500           []
24290   AUA93J_LOWW_LOWK_20170901210500           []
24291  AUA929K_LOWW_LOWK_20170901181500           []
24293  GDK01CA_LOWW_LOWK_20170901100000           []
28912  AHY9223_UBBB_UGTB_20170901043000           []
29026   TGZ785_UGTB_ULLI_20170901172000           []
29383   TGZ786_ULLI_UGTB_20170901214500           []
29614   SBI955_UUDD_UGTB_20170901105500           []

# 16
Append "flights2" to "clean".
<br>Sort the complete dataframe accoridng to the column 'flight'.
<br>Reset the index.

In [18]:
clean5 = clean5.append(flights2)
clean5.sort_values(by=['flight'], inplace=True)
clean5.reset_index(drop=True, inplace=True)
clean5

flight  \
0        A6RTS_OMDW_EGGW_20170901055500   
1       AAB249_EBKT_LFMN_20170901070500   
2       AAB249_LFMN_LIEO_20170901090500   
3       AAB25C_LOWW_UUMO_20170901143500   
4       AAB571_EBBR_EGNV_20170901050500   
...                                 ...   
29912   XLF056_LFPG_MMUN_20170901153100   
29913  XLF1451_LFML_LFPG_20170901030500   
29914   XRO546_LFMD_EGCC_20170901180000   
29915   XRO552_EHAM_LFMD_20170901150000   
29916    Z3MKD_LWSK_EDDF_20170901130000   

                                            sector_hours  
0      [LTAAAEU67_0800_0900_E, LTAAAEU67_0900_1000_E,...  
1      [EBBUWLS_0620_0700_E, EBBUWLS_0700_0800_E, EBB...  
2      [LFMMMNST_0800_0900_E, LFMMMNST_0900_1000_E, L...  
3      [LZBBL33_1400_1500_E, LZBBL33_1500_1600_E, LZB...  
4      [EBBUNWC_0530_0600_E, EBBUNWC_0600_0620_E, EBB...  
...                                                  ...  
29912  [LFFFTH_1500_1600_E, LFFFTH_1600_1700_E, LFFFU...  
29913  [LFMMMALY_0330_0400_E, LFMMMALY_0400_0430_E, L...  
29914  [LFMMSBAM_1715_1800_E, LFMMSBAM_1800_1900_E, L...  
29915  [EHAACOD_1500_1600_E, EHAACOD_1600_1700_E, EHA...  
29916  [LWSSLOW_1200_1300_E, LWSSLOW_1300_1400_E, LYB...  

[29917 rows x 2 columns]

# 17
Export the dataframe as pickle (in order to preserve the type of the objects).

In [19]:
clean5.to_pickle('S1_flights_sectorhours.pkl')

# 18    
    
    Con l'istante di partenza fissato, era necessario che:
    1) entry - start >= 0   ovvero:   start - entry <= 0     ovvero: start <= entry
    2) end - entry >= 0     ovvero:   end >= entry           ovvero: entry <= end
    
    Con la flessibilità di più/meno 30 minuti, basta che:
    1) (entry + 30) - start >= 0     ovvero: start - 30 <= entry     ovvero: start <= entry + 30
    2) end - (entry - 30) >= 0       ovvero: end + 30 > entry        ovvero: entry <= end + 30
    
    -----
    Esempio:

    A) settore-ora attivo dal min 60 al min 120
    B) settore-ora attivo dal min 0 al min 60
    C) settore-ora attivo dal min 100 al min 160

    Volo deve transitare per il settore al min 80 (fissato):
    Settore-ora A) OK
    Settore-ora B) NO
    Settore-ora C) NO
    
    Anticipando di 30 minuti (min 50):
    Settore-ora A) NO
    Settore-ora B) OK
    Settore-ora C) NO
    
    Posticipando di 30 minuti (min 110):
    Settore-ora A) OK
    Settore-ora B) NO
    Settore-ora C) OK